cdap-app-fabric/src/main/java/co/cask/cdap/common/twill/TwillAppLifecycleEventHandler.java

```diff
@@ -0,0 +1,277 @@
+import org.apache.twill.api.RunId;
+import org.apache.twill.zookeeper.ZKClientService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+
+/**
+ * A Twill {@link EventHandler} that responds to Twill application lifecycle events and aborts the application if
+ * it cannot provision container for some runnable.
+ */
+public class TwillAppLifecycleEventHandler extends EventHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(TwillAppLifecycleEventHandler.class);
+  private static final Gson GSON = new Gson();
+  private static final String HADOOP_CONF_FILE_NAME = "hConf.xml";
+  private static final String CDAP_CONF_FILE_NAME = "cConf.xml";
+
+  private long abortTime;
+  private boolean abortIfNotFull;
+  private RunId twillRunId;
+  private ProgramRunId programRunId;
+  private ProgramStateWriter programStateWriter;
+  private ZKClientService zkClientService;
+  private AtomicBoolean launchedContainer;
+  private AtomicBoolean stoppedContainer;
+
+  /**
+   * Constructs an instance of TwillAppLifecycleEventHandler that abort the application if some runnable has no
+   * containers, same as calling
+   * {@link #TwillAppLifecycleEventHandler(long, boolean, ProgramOptions)} with second
+   * parameter as {@code false}.
+   *
+   * @param abortTime Time in milliseconds to pass before aborting the application if no container is given to
+   *                  a runnable.
+   */
+  public TwillAppLifecycleEventHandler(long abortTime) {
+    this(abortTime, false, null);
+  }
+
+  /**
+   * Constructs an instance of TwillAppLifecycleEventHandler that abort the application if some runnable has no
+   * containers and a program state writer to respond to Twill application lifecycle events
+   *
+   * @param abortTime Time in milliseconds to pass before aborting the application if no container is given to
+   *                  a runnable.
+   * @param options the program options, if the event handler was attached to monitor program states
+   */
+  public TwillAppLifecycleEventHandler(long abortTime, @Nullable ProgramOptions options) {
+    this(abortTime, false, options);
+  }
+
+  /**
+   * Constructs an instance of TwillAppLifecycleEventHandler that abort the application if some runnable has not enough
+   * containers.
+   * @param abortTime Time in milliseconds to pass before aborting the application if no container is given to
+   *                  a runnable.
+   * @param abortIfNotFull If {@code true}, it will abort the application if any runnable doesn't meet the expected
+   *                       number of instances.
+   * @param options the program options, if the event handler was attached to monitor program states
+   */
+  public TwillAppLifecycleEventHandler(long abortTime, boolean abortIfNotFull, @Nullable ProgramOptions options) {
```

Comment:
For the program AM, it always take a `ProgramOptions`, right? I think it's better to keep the existing `AbortOnTimeoutEventHandler` and have the `TwillProgramEventHandler` extends from it. In general, try to minimize usage of `null`.